### PR TITLE
[GCS]Use new getting all available resources interface instead of pub-sub …

### PR DESCRIPTION
**Pull request by WangTaoTheTonic**
…in state.py

## Why are these changes needed?

now in state.py we use pub-sub to get available resources of cluster. This relies on broadcast message of HEARTBEAT pattern, which will be triggered every 100ms(default). This means we will cost ~100 ms, 3 * 100ms at most(if one client report heartbeat abnormaly), to get the returned value, which is too expensive and not necessary.

This PR added a new interface in gcs to get all available resources directly, reuse internal data structure. This will largely improve the performance as it need only a getting rpc.

## Related issue number

Part of https://github.com/ray-project/ray/issues/10355

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/gcs/gcs_client/test/service_based_gcs_client_test.cc [hunk 1]
```diff
@@ -324,6 +324,20 @@ class ServiceBasedGcsClientTest : public ::testing::Test {
     return WaitReady(promise.get_future(), timeout_ms_);
   }
 
+  std::vector<rpc::AvailableResources> GetAllAvailableResources() {
+    std::promise<bool> promise;
+    std::vector<rpc::AvailableResources> resources;
+    RAY_CHECK_OK(gcs_client_->Nodes().AsyncGetAllAvailableResources(
+        [&resources, &promise](Status status,
+                               const std::vector<rpc::AvailableResources> &result) {
+          assert(!result.empty());
```
Comment (ffbin) [inline src/ray/gcs/gcs_client/test/service_based_gcs_client_test.cc:333]: EXPECT_TRUE(!result.empty());

File: src/ray/protobuf/gcs.proto [hunk 2]
```diff
@@ -222,6 +222,13 @@ message ResourceTableData {
   double resource_capacity = 1;
 }
 
+message AvailableResources {
+  // Node manager client id
```
Comment (ffbin) [inline src/ray/protobuf/gcs.proto:226]:
```suggestion
  // Node manager client id.
```

File: src/ray/protobuf/gcs.proto [hunk 3]
```diff
@@ -222,6 +222,13 @@ message ResourceTableData {
...
+  bytes client_id = 1;
```
Comment (ffbin) [inline src/ray/protobuf/gcs.proto:227]: node_id ?

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 4]
```diff
@@ -635,6 +635,24 @@ Status ServiceBasedNodeInfoAccessor::AsyncGetResources(
...
 }
 
+Status ServiceBasedNodeInfoAccessor::AsyncGetAllAvailableResources(
+    const MultiItemCallback<rpc::AvailableResources> &callback) {
+  rpc::GetAllAvailableResourcesRequest request;
+  client_impl_->GetGcsRpcClient().GetAllAvailableResources(
+      request,
+      [callback](const Status &status, const rpc::GetAllAvailableResourcesReply &reply) {
+        std::vector<rpc::AvailableResources> result;
+        result.reserve((reply.resources_list_size()));
```
Comment (ffbin) [inline src/ray/gcs/gcs_client/service_based_accessor.cc:645]: use `VectorFromProtobuf` function directly

File: src/ray/gcs/gcs_client/test/service_based_gcs_client_test.cc [hunk 5]
```diff
@@ -706,6 +720,36 @@ TEST_F(ServiceBasedGcsClientTest, TestNodeHeartbeat) {
...
+  auto node_info = Mocker::GenNodeInfo();
+  RAY_CHECK(RegisterNode(*node_info));
+
+  // Report heartbeat of a node to GCS.
+  ClientID node_id = ClientID::FromBinary(node_info->node_id());
+  auto heartbeat = std::make_shared<rpc::HeartbeatTableData>();
+  heartbeat->set_client_id(node_id.Binary());
+  // Set this flag because GCS won't publish unchanged heartbeat.
+  heartbeat->set_should_global_gc(true);
+  (*heartbeat->mutable_resources_available())["CPU"] = 1.0;
```
Comment (rkooo567) [inline src/ray/gcs/gcs_client/test/service_based_gcs_client_test.cc:742]: Can you add one more entry? Like "GPU": 10.0

File: src/ray/gcs/gcs_server/gcs_node_manager.cc [hunk 6]
```diff
@@ -354,6 +354,21 @@ void GcsNodeManager::HandleGetInternalConfig(const rpc::GetInternalConfigRequest
       gcs_table_storage_->InternalConfigTable().Get(UniqueID::Nil(), get_system_config));
 }
 
+void GcsNodeManager::HandleGetAllAvailableResources(
+    const rpc::GetAllAvailableResourcesRequest &request,
+    rpc::GetAllAvailableResourcesReply *reply,
+    rpc::SendReplyCallback send_reply_callback) {
+  for (auto iter : GetClusterRealtimeResources()) {
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_node_manager.cc:361]: `const auto &iter`

**Review by rkooo567 — COMMENTED** [review]
LGTM. Just a couple minor comments!

File: src/ray/gcs/gcs_client/test/service_based_gcs_client_test.cc (hunk 1 above)
Comment (WangTaoTheTonic) [reply src/ray/gcs/gcs_client/test/service_based_gcs_client_test.cc:333]: OK

File: src/ray/protobuf/gcs.proto (hunk 2 above)
Comment (WangTaoTheTonic) [reply src/ray/protobuf/gcs.proto:226]: ok

File: src/ray/protobuf/gcs.proto (hunk 3 above)
Comment (WangTaoTheTonic) [reply src/ray/protobuf/gcs.proto:227]: nice catch!

File: src/ray/gcs/gcs_client/service_based_accessor.cc (hunk 4 above)
Comment (WangTaoTheTonic) [reply src/ray/gcs/gcs_client/service_based_accessor.cc:645]: ok

File: python/ray/state.py [hunk 7]
```diff
@@ -770,48 +770,33 @@ def available_resources(self):
...
-            if (raw_message is None or raw_message["pattern"] !=
-                    gcs_utils.XRAY_HEARTBEAT_PATTERN):
-                continue
-            data = raw_message["data"]
-            pub_message = gcs_utils.PubSubMessage.FromString(data)
-            heartbeat_data = pub_message.data
-            message = gcs_utils.HeartbeatTableData.FromString(heartbeat_data)
+        all_available_resources = \
+            self.global_state_accessor.get_all_available_resources()
+        for i in range(len(all_available_resources)):
```
[the remark this answers is not in the document]
Comment (WangTaoTheTonic) [reply python/ray/state.py:775]: ok!

File: python/ray/state.py [hunk 8]
```diff
@@ -770,48 +770,33 @@ def available_resources(self):
...
 
-            # Update clients in cluster
-            client_ids = self._live_client_ids()
-
-            # Remove disconnected clients
-            for client_id in list(available_resources_by_id.keys()):
-                if client_id not in client_ids:
-                    del available_resources_by_id[client_id]
+        # Update clients in cluster
+        client_ids = self._live_client_ids()
```
[the remark this answers is not in the document]
Comment (WangTaoTheTonic) [reply python/ray/state.py:788]: all right~

**Comment by WangTaoTheTonic** [comment]
Comments responsed and master codes merged.

There comes a new API `_available_resources_per_node` who uses heartbeat publish.
- [ ] we need change this.

**Comment by rkooo567** [comment]
There's a merge conflict! Please remove the author action required tag once it is ready. Also, Windows CI should be green now :) (we blacklisted all failing tests)

File: src/ray/gcs/gcs_client/test/service_based_gcs_client_test.cc (hunk 5 above)
Comment (WangTaoTheTonic) [reply src/ray/gcs/gcs_client/test/service_based_gcs_client_test.cc:742]: ok

File: src/ray/gcs/gcs_server/gcs_node_manager.cc (hunk 6 above)
Comment (WangTaoTheTonic) [reply src/ray/gcs/gcs_server/gcs_node_manager.cc]: ok

**Comment by WangTaoTheTonic** [comment]
Some tests related with `ray.options` failed. Could not get the reason. Still digging...

**Review by rkooo567 — APPROVED** [review]

**Pull request merged**